### PR TITLE
Create concatenate3, a more efficient version of rec_concatenate

### LINKED ISSUE #243: rec_concatenate slows down ghosting computations
In [this notebook](http://nbviewer.ipython.org/gist/mrocklin/dc7fd426a6e405fef275) we find that concatentation within dask.array.ghost is a bottleneck.  Currently this depends on `rec_concatenate` which makes several recursive calls to `numpy.concatenate`; this is wasteful. 

Instead it would be nice to allocate one target array and then insert all of the source arrays into the proper place

```python
result = np.empty(...)  # allocate once

for location, x in ... : # some recursive looping, possibly with itertools.product
    result[*location] = x  # insert each array individually
```

It would be interesting to see how much this can improve performance.  It would also be good to profile the application in the above notebook to establish that the lower bound is good enough to warrant this effort.

cc @cowlicks @ahmadia @ketch

### PR DESCRIPTION
**Pull request by mrocklin**
Fixes #243 

This provides a procedural nd-concatenate function that only results in one copy, rather than n copies for n the number of dimensions.

### CODE REVIEW

**Comment by mrocklin** [comment]
This cuts the concatenate overhead in @ketch 's notebook down by a factor of two.

**Comment by mrocklin** [comment]
@cowlicks you're probably the best person to review this if you get a chance.

File: dask/array/core.py [hunk 1]
```diff
@@ -1777,3 +1777,63 @@ def bincount(x, weights=None, minlength=None):
...
+    """
+    if not isinstance(seq, list):
+        return seq
+    else:
+        return deepfirst(seq[0])
+
+
+def concatenate3(arrays):
+    """ Recursive np.concatenate
+
```
Comment (shoyer) [inline dask/array/core.py]: I would just note that the arguments need to be provided as a recursive list of lists

**Comment by shoyer** [comment]
Generally looks pretty good to me. I can understand this one much better than the original `rec_concatenate` with `axis` argument.

File: dask/array/core.py (hunk 1 above)
Comment (mrocklin) [inline dask/array/core.py]: Fixed

**Comment by mrocklin** [comment]
I've gone ahead and removed `rec_concatenate`, replacing with `concatenate3` in all cases.

**Comment by mrocklin** [comment]
Any more thoughts?

**Comment by cowlicks** [comment]
Can you add a 3d and 1d test?

**Comment by mrocklin** [comment]
Good idea.  Added

**Comment by mrocklin** [comment]
Further thoughts?

**Comment by mrocklin** [comment]
merging soon if no comment

**Comment by cowlicks** [comment]
LGTM

**Pull request merged**